tests/lib/prepare.sh

```diff
@@ -0,0 +1,141 @@
+        echo "$want_pw" > /tmp/new-shadow
+        tail -n +2 /etc/shadow >> /tmp/new-shadow
+        cp -v /tmp/new-shadow $UNPACKD/etc/shadow
+        
+        # we need the test user in the image
+        chroot $UNPACKD adduser --quiet --no-create-home --disabled-password --gecos '' test
+
+        # modify sshd so that we can connect as root
+        sed -i 's/\(PermitRootLogin\|PasswordAuthentication\)\>.*/\1 yes/' $UNPACKD/etc/ssh/sshd_config
+
+        # FIXME: install would be better but we don't have dpkg on
+        #        the image
+        # unpack our freshly build snapd into the new core snap
+        dpkg-deb -x ${SPREAD_PATH}/../snapd_*.deb $UNPACKD
+        
+        # build new core snap for the image
+        snapbuild $UNPACKD $IMAGE_HOME
+
+        # FIXME: how to test store updated of ubuntu-core with that?
+        
+        # create new image with the modified ubuntu-core snap
+        IMAGE=all-snap-amd64.img
+        /snap/bin/ubuntu-device-flash core 16 --channel edge --gadget pc  --kernel pc-kernel --os $IMAGE_HOME/ubuntu-core_*.snap --install snapweb  --output $IMAGE_HOME/$IMAGE
+        
+        # mount fresh image and add all our SPREAD_PROJECT data
+        kpartx -avs $IMAGE_HOME/$IMAGE
+        # FIXME: hardcoded mapper location, parse from kpartx
+        mount /dev/mapper/loop2p3 /mnt
+        mkdir -p /mnt/user-data/
+        cp -avr /home/gopath /mnt/user-data/
+
+        # create test user home dir
+        mkdir -p /mnt/user-data/test
+        chown 1001:1001 /mnt/user-data/test
+        
+        # FIXUP silly systemd
+        mkdir -p /mnt/system-data/etc/systemd/system/snapd.service.d
+        cat <<EOF > /mnt/system-data/etc/systemd/system/snapd.service.d/local.conf
+[Unit]
+StartLimitInterval=0
+[Service]
+Environment=SNAPD_DEBUG_HTTP=7 SNAP_REEXEC=0
+EOF
+    
+        umount /mnt
+        kpartx -d  $IMAGE_HOME/$IMAGE
+    
+        # the reflash magic
+        # FIXME: ideally in initrd, but this is good enough for now
+        cat > $IMAGE_HOME/reflash.sh << EOF
+#!/bin/sh -ex
+mount -t tmpfs none /tmp
+cp /bin/busybox /tmp
+# blow away everything
+/tmp/busybox dd if=$IMAGE_HOME/$IMAGE of=/dev/sda bs=4M
+# and reboot
+/tmp/busybox sync
+/tmp/busybox echo b > /proc/sysrq-trigger
+EOF
+        chmod +x $IMAGE_HOME/reflash.sh
+
+        # FIXME: hardcoded sda1
+        cat >/boot/grub/grub.cfg <<EOF
+set default=0
+set timeout=2
+menuentry 'flash-all-snaps' {
+linux /vmlinuz root=/dev/sda1 ro init=$IMAGE_HOME/reflash.sh console=ttyS0
+initrd /initrd.img
+}
+EOF
+
+        # Reboot !
+REBOOT
+    fi
+
+    # verify after the first reboot that we are now in the all-snap world
+    if [ $SPREAD_REBOOT = 1 ]; then
+        echo "Ensure we are now in an all-snap world"
+        if [ -e /var/lib/dpkg/status ]; then
+            echo "Rebooting into all-snap system did not work"
+            exit 1
+        fi
+        # Snapshot the state.json
+        if [ ! -f $SPREAD_PATH/snapd-state.tar.gz ]; then
```

Comment:
When reusing, next time we reboot we'll be in all snap world, and will have that snapshot at hand, but $SPREAD_REBOOT will be `0`. I think we want to move all of this REBOOT=1 logic up, before the REBOOT=0, and also base the decision to skip the pre-reboot logic if we are already in the all-snap land with a saved state.

Then, return at the end of those conditions, so we can actually de-indent all the REBOOT=0 and have it til the end.